File: app/src/main/java/com/aegamesi/steamtrade/libs/GlideImageGetter.java
package com.aegamesi.steamtrade.libs;

/*
https://github.com/bumptech/glide/issues/3328
 */

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.Html;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.Request;
import com.bumptech.glide.request.RequestOptions;
import com.bumptech.glide.request.target.SizeReadyCallback;
import com.bumptech.glide.request.target.Target;
import com.bumptech.glide.request.transition.Transition;

//ImageGetter for chat emoticons.
public class GlideImageGetter implements Html.ImageGetter {
    private Context context;
    private TextView textView;
    private float pixelsToDp;

    public GlideImageGetter(View t, Context context) {
        this.context = context;
        this.textView = (TextView) t;

        Resources resources = context.getResources();
        DisplayMetrics metrics = resources.getDisplayMetrics();
        pixelsToDp = ((float) metrics.densityDpi / DisplayMetrics.DENSITY_DEFAULT);
    }

    @Override
    public Drawable getDrawable(String source) {
        Log.i("GlideImageGetter", "Loading: " + source);
        int size = (int) (16.0f * pixelsToDp);

        BitmapDrawablePlaceholder drawable = new BitmapDrawablePlaceholder();

        Glide.with(context)
                .asBitmap()
                .load(source)
                .apply(new RequestOptions().override(size, size))
                .into(drawable);

        return drawable;
    }

    private class BitmapDrawablePlaceholder extends BitmapDrawable implements Target<Bitmap> {

        protected Drawable drawable;

        BitmapDrawablePlaceholder() {
            super(context.getResources(), Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888));
        }

        @Override
        public void draw(final Canvas canvas) {
            if (drawable != null) {
                drawable.draw(canvas);
            }
        }

        private void setDrawable(Drawable drawable) {
            this.drawable = drawable;
            int drawableWidth = drawable.getIntrinsicWidth();
            int drawableHeight = drawable.getIntrinsicHeight();
            int maxWidth = textView.getMeasuredWidth();
            if (drawableWidth > maxWidth) {
                int calculatedHeight = maxWidth * drawableHeight / drawableWidth;
                drawable.setBounds(0, 0, maxWidth, calculatedHeight);
                setBounds(0, 0, maxWidth, calculatedHeight);
            } else {
                drawable.setBounds(0, 0, drawableWidth, drawableHeight);
                setBounds(0, 0, drawableWidth, drawableHeight);
            }

            textView.setText(textView.getText());
        }

        @Override
        public void onLoadStarted(@Nullable Drawable placeholderDrawable) {
            if(placeholderDrawable != null) {
                setDrawable(placeholderDrawable);
            }
        }

        @Override
        public void onLoadFailed(@Nullable Drawable errorDrawable) {
            if (errorDrawable != null) {
                setDrawable(errorDrawable);
            }
        }

        @Override
        public void onResourceReady(@NonNull Bitmap bitmap, @Nullable Transition<? super Bitmap> transition) {
            setDrawable(new BitmapDrawable(context.getResources(), bitmap));
        }

        @Override
        public void onLoadCleared(@Nullable Drawable placeholderDrawable) {
            if(placeholderDrawable != null) {
                setDrawable(placeholderDrawable);
            }
        }

        @Override
        public void getSize(@NonNull SizeReadyCallback cb) {
            textView.post(() -> cb.onSizeReady(textView.getWidth(), textView.getHeight()));
        }

        @Override
        public void removeCallback(@NonNull SizeReadyCallback cb) {}

        @Override
        public void setRequest(@Nullable Request request) {}

        @Nullable
        @Override
        public Request getRequest() {
            return null;
        }

        @Override
        public void onStart() {}

        @Override
        public void onStop() {}

        @Override
        public void onDestroy() {}

    }
}

File: app/src/main/java/com/aegamesi/steamtrade/libs/GlideImageGetter.kt
package com.aegamesi.steamtrade.libs

/*
https://github.com/bumptech/glide/issues/3328
 */

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.text.Html
import android.util.DisplayMetrics
import android.view.View
import android.widget.TextView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.Request
import com.bumptech.glide.request.RequestOptions
import com.bumptech.glide.request.target.SizeReadyCallback
import com.bumptech.glide.request.target.Target
import com.bumptech.glide.request.transition.Transition

//ImageGetter for chat emoticons.
class GlideImageGetter(t: View, private val context: Context) : Html.ImageGetter {
    private val textView: TextView = t as TextView
    private val pixelsToDp: Float

    init {
        val resources = context.resources
        val metrics = resources.displayMetrics
        pixelsToDp = metrics.densityDpi.toFloat() / DisplayMetrics.DENSITY_DEFAULT
    }

    override fun getDrawable(source: String): Drawable {

        //Let's make the emoticons a bit larger. Was: 16.0f
        val size = (24.0f * pixelsToDp).toInt()
        val drawable = BitmapDrawablePlaceholder()

        Glide.with(context)
                .asBitmap()
                .load(source)
                .apply(RequestOptions().override(size, size))
                .into(drawable)

        return drawable
    }

    private inner class BitmapDrawablePlaceholder internal constructor() : BitmapDrawable(context.resources, Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888)), Target<Bitmap> {

        var bitMapDrawable: Drawable? = null

        override fun draw(canvas: Canvas) {
            if (bitMapDrawable != null) {
                bitMapDrawable!!.draw(canvas)
            }
        }

        private fun setDrawable(drawable: Drawable) {
            this.bitMapDrawable = drawable
            val drawableWidth = drawable.intrinsicWidth
            val drawableHeight = drawable.intrinsicHeight
            val maxWidth = textView.measuredWidth
            if (drawableWidth > maxWidth) {
                val calculatedHeight = maxWidth * drawableHeight / drawableWidth
                drawable.setBounds(0, 0, maxWidth, calculatedHeight)
                setBounds(0, 0, maxWidth, calculatedHeight)
            } else {
                drawable.setBounds(0, 0, drawableWidth, drawableHeight)
                setBounds(0, 0, drawableWidth, drawableHeight)
            }

            textView.text = textView.text
        }

        override fun onLoadStarted(placeholderDrawable: Drawable?) {
            if (placeholderDrawable != null) {
                setDrawable(placeholderDrawable)
            }
        }

        override fun onLoadFailed(errorDrawable: Drawable?) {
            if (errorDrawable != null) {
                setDrawable(errorDrawable)
            }
        }

        override fun onResourceReady(bitmap: Bitmap, transition: Transition<in Bitmap>?) {
            setDrawable(BitmapDrawable(context.resources, bitmap))
        }

        override fun onLoadCleared(placeholderDrawable: Drawable?) {
            if (placeholderDrawable != null) {
                setDrawable(placeholderDrawable)
            }
        }

        override fun getSize(cb: SizeReadyCallback) {
            textView.post { cb.onSizeReady(textView.width, textView.height) }
        }

        override fun removeCallback(cb: SizeReadyCallback) {}

        override fun setRequest(request: Request?) {}

        override fun getRequest(): Request? {
            return null
        }

        override fun onStart() {}

        override fun onStop() {}

        override fun onDestroy() {}

    }
}